src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,48 @@
 package com.es.phoneshop.model.product;
 
+
+import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private List<Product> productList = new ArrayList<>();
+
+    public List<Product> getProductList() {
+        return productList;
+    }
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        return productList.stream().filter(product -> product.getId().equals(id)).findAny()
+                .orElseThrow(() -> new IllegalArgumentException("Such object doesn't exsist"));
+
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        return productList.stream()
+                .filter(product -> product.getPrice() != null && product.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        if (productList.stream().anyMatch(prod -> prod.getId().equals(product.getId()))) {
+            throw new IllegalArgumentException("Object's id isn't unic");
+        } else {
+            productList.add(product);
+        }
     }
 
     @Override
     public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+
+        if (productList.stream().anyMatch(product -> product.getId().equals(id))) {
+            Product delProd = this.getProduct(id);
```

Comment:
Nice idea to make use of getProduct method in here, but why do you use this if-else-throw logic? If the product is not found, the getProduct method will throw the same exception